### DISCUSSION TITLE
Filtering noise from unused INMP441 channel

### DISCUSSION THREAD

**Discussion by cmcooper1971** [pschatzmann/arduino-audio-tools#1247] [Q&A] [answered]
Hi All,

I have successfully built a I2S Digital Amplifier, which consists of an SD card reader, an I2S Microphone (INMP441) linked to an I2S Texas Instruments Class D Amplifier (TAS5760L), all of which is controlled using Phils, Arduino Audio Tools library.

Please see previous post https://github.com/pschatzmann/arduino-audio-tools/discussions/1229.

Following on from that post, I am trying to apply effects as per Phil's response, but the unused channel (as I am only using one microphone input) is causing noise as per Phil's WiKi https://github.com/pschatzmann/arduino-audio-tools/wiki/External-ADC#inmp441---mems-microphone

Therefore my thought was to convert the input from 2 channels to 1, then applying Audio Effects. However when I do, I still get the noise problem reference in the WiKi.

The WiKi states however to try ConverterFillLeftAndRight, but I do not understand how to implement this and cannot find any examples accordingly.

I have attempted to convert 2 channels to 1 using the following code:

```
/*-----------------------------------------------------------------*/

uint8_t from_channels = 2;
uint8_t to_channels = 1;

AudioInfo info(44100, 2, 16);

I2SStream in;

ChannelFormatConverterStreamT<int16_t> conv(in);
AudioEffectStream effects(conv);

I2SStream out;

StreamCopy mic(out, effects);

PitchShift pitchShift(1.5);

/*-----------------------------------------------------------------*/
```

```
/*-----------------------------------------------------------------*/

// Enable MIC & disable Player

void enableMIC() {

	// Setup I2S in

	AudioLogger::instance().begin(Serial, AudioLogger::Warning);

	debugln("starting I2S...");
	auto config_in = in.defaultConfig(RX_MODE);
	config_in.copyFrom(info);
	config_in.i2s_format = I2S_STD_FORMAT;
	config_in.is_master = true;
	config_in.port_no = 0;
	config_in.pin_bck = I2S_BCLK_0;
	config_in.pin_ws = I2S_LRC_0;
	config_in.pin_data = I2S_DIN_0;
	
	conv.begin(from_channels, to_channels);

	effects.addEffect(pitchShift);
	in.begin(config_in);
	effects.begin(config_in);
	
	// Setup I2S out

	auto config_out = out.defaultConfig(TX_MODE);
	config_out.copyFrom(info);
	config_out.i2s_format = I2S_STD_FORMAT;
	config_out.is_master = true;
	config_out.port_no = 1;
	config_out.pin_bck = I2S_BCLK_1;
	config_out.pin_ws = I2S_LRC_1;
	config_out.pin_data = I2S_DOUT_1;
	config_out.pin_mck = I2S_MCLK_1;
	out.begin(config_out);

	mic.begin(out, effects);

	debugln("Mic Enabled");
	playerControlFlagChanged = LOW;

}	// Close function

/*-----------------------------------------------------------------*/
```

Regards,

Christopher

**Comment by pschatzmann** [comment]
You can use the [converter object](https://pschatzmann.github.io/arduino-audio-tools/classaudio__tools_1_1_converter_fill_left_and_right.html) as parameter when you call [copy](https://pschatzmann.github.io/arduino-audio-tools/classaudio__tools_1_1_stream_copy.html#a226bbd7cd5d674fabacfb94711178728) or you can build your own output chain by adding a [ConverterStream](https://pschatzmann.github.io/arduino-audio-tools/classaudio__tools_1_1_converter_stream.html)

Dont forget to specify in the constructor which channel should get overwritten (LeftIsEmpty, RightIsEmpty)!

Alternatively you can read the data with 32 bits and then convert it to 16...

**Reply by cmcooper1971** [reply]
Hi Phil,

Thanks for replying.

Could you confirm that I am on the right track with this please?

```
AudioInfo info(44100, 2, 16);

I2SStream in;

ConverterStream<int16_t, ConverterFillLeftAndRight<int16_t>> RightIsEmpty;

AudioEffectStream effects(in);

I2SStream out;

StreamCopy mic(out, effects);

PitchShift pitchShift(1.5);
ADSRGain adsr(0.0001, 0.0001, 0.9, 0.0002);
```

Through your examples I am struggling to understand how to call/apply this within the setup or loop however.

Thanks as always for your library and help.

**Comment by pschatzmann** [comment, marked as answer]
```
I2SStream in;
ConverterFillLeftAndRight<int16_t> left_and_right(RightIsEmpty);
ConverterStream<int16_t> converter(in, left_and_right);
AudioEffectStream effects(converter);
```
But [dont forget to check on which channel you have the issue](https://github.com/pschatzmann/arduino-audio-tools/wiki/External-ADC)...

**Reply by cmcooper1971** [reply]
That makes sense, thank you. 

I do get an error on the ConverterStream line:

`ConverterStream<int16_t> converter(in, left_and_right);`

When building the sketch, the error log states:

```
I2S_Receiver_Unit_Design.ino: 124:24: error: wrong number of template arguments (1, should be 2)
   ConverterStream<int16_t> converter(in, left_and_right)
```

Reviewing the definition, the line does include both parameters, do I cannot see why it throws the error?

```
    public:
        ConverterStream(Stream &stream, ConverterT &converter) : AudioStream() {
            p_converter = &converter;
            p_stream = &stream;
        }
```

**Reply by pschatzmann** [reply]
I can't reproduce your error with the latest version of my library.
Maybe you use a old version ?

**Reply by cmcooper1971** [reply]
Well that will teach me a lesson Phil.

Thank you once again.

**Reply by cmcooper1971** [reply]
Hi Phil,

Everything is working, but I thought you might like to know that the only solution I could get working in the end was to connect two INMP441 microphones.

As soon as I did this, the noise problem you reference in your WiKi didn't occur when applying various effects, such as pitch shift. 

To confirm, as per your previous note, I have checked which channel is creating the noise and toggled my circuit to right and left, to prove my design.

I have also applied the filter to the input and moved the effect to the output, which didn't produce any differences. Finally, I deployed the effect without the converter and as long as I use two microphones, it works.

If you have any other suggestions, they would be appreciated.

Regards,

Christopher

```
/*-----------------------------------------------------------------*/

AudioInfo info(44100, 2, 16);

I2SStream in;
I2SStream out;

ConverterFillLeftAndRight<int16_t> left_and_right(RightIsEmpty);
ConverterStream<int16_t> converter(in, left_and_right);

AudioEffectStream effects(converter);

StreamCopy mic(out, effects);

PitchShift pitchShift(1.2);

/*-----------------------------------------------------------------*/
```

My function:

```
/*-----------------------------------------------------------------*/

// Enable MIC & disable Player

void enableMIC() {

	decoder.end();
	mp3.end();

	// Start I2S in

	AudioLogger::instance().begin(Serial, AudioLogger::Warning);

	debugln("starting I2S...");
	auto config_in = in.defaultConfig(RX_MODE);
	config_in.copyFrom(info);
	config_in.i2s_format = I2S_STD_FORMAT;
	config_in.is_master = true;
	config_in.port_no = 0;
	config_in.pin_bck = I2S_BCLK_0;
	config_in.pin_ws = I2S_LRC_0;
	config_in.pin_data = I2S_DIN_0;

	effects.addEffect(pitchShift);
	in.begin(config_in);
	effects.begin(config_in);

	//in.begin(config_in);

	// Start I2S out

	auto config_out = out.defaultConfig(TX_MODE);
	config_out.copyFrom(info);
	config_out.i2s_format = I2S_STD_FORMAT;
	config_out.is_master = true;
	config_out.port_no = 1;
	config_out.pin_bck = I2S_BCLK_1;
	config_out.pin_ws = I2S_LRC_1;
	config_out.pin_data = I2S_DOUT_1;
	config_out.pin_mck = I2S_MCLK_1;
	out.begin(config_out);

	mic.begin(out, effects);

	debugln("Mic Enabled");
	digitalWrite(whiteLED, LOW);
	playerControlFlagChanged = LOW;

	displayVolumeRegister();

}	// Close function

/*-----------------------------------------------------------------*/
```

**Comment by pschatzmann** [comment]
There was a bug on the reading side of the ConverterStream. I have committed the related correction

**Reply by cmcooper1971** [reply]
Hi Phil.

Thanks for looking at that. I was working to much on the problem, as I had an epiphany this morning, in that I did not include the converter.begin() in my function, see above.

Trying this morning, I cannot seem to figure out the right way to call it, or in fact do I need to?

Thanks.

**Reply by pschatzmann** [reply]
That's what I was thinking first as well, but currently calling begin does nothing, though it is good practice to call begin().
I think the root cause was really that the readBytes function was not doing anything because the result variable was not initialized. 

ps. don't forget to update (or install) the new version of the library to get the correction.

**Reply by cmcooper1971** [reply]
Hi Phil,

Although on first upload of your latest library, it didn't work, as I still received noise when I reverted to one microphone, I decided to move the effects to the output of the microphone function, leaving the channel converter on the input.

This worked!

Many many thanks, I am thrilled. I can now design my receiver PCB.

For reference this is my test code as working:

```
/*-----------------------------------------------------------------*/

// Audio object

AudioInfo info(44100, 2, 16);

// Audio streams in and out

I2SStream in;
I2SStream out;

ConverterFillLeftAndRight<int16_t> left_and_right(RightIsEmpty);
ConverterStream<int16_t> converter(in, left_and_right);

AudioEffectStream effects(out);

StreamCopy mic(effects, converter);

// Effect variables

int16_t clipThreashold = 3000;
int16_t setMaxInput = 6500;
float setfuzzEffectValue = 6.5;
int16_t setfuzzMaxValue = 300;

int16_t setDuration = 750;
float setDepth = 0.5;
float setFeedback = 1.0;
int32_t setSampleRate = 44100;

// Effects

PitchShift pitchShift(1.5);
Distortion distortion(clipThreashold, setMaxInput);
Fuzz fuzz(setfuzzEffectValue, setfuzzMaxValue);
Delay inDelay(setDuration, setDepth, setFeedback, setSampleRate, false);

/*-----------------------------------------------------------------*/
```

My function:

```
/*-----------------------------------------------------------------*/

// Enable MIC & disable Player

void enableMIC() {

	// Stop player

	decoder.end();
	mp3.end();

	// Apply chosen effect(s)

	effects.addEffect(pitchShift);
	//effects.addEffect(distortion);
	//effects.addEffect(fuzz);
	//effects.addEffect(inDelay);

	// Start I2S in

	AudioLogger::instance().begin(Serial, AudioLogger::Warning);

	debugln("starting I2S...");
	auto config_in = in.defaultConfig(RX_MODE);
	config_in.copyFrom(info);
	config_in.i2s_format = I2S_STD_FORMAT;
	config_in.is_master = true;
	config_in.port_no = 0;
	config_in.pin_bck = I2S_BCLK_0;
	config_in.pin_ws = I2S_LRC_0;
	config_in.pin_data = I2S_DIN_0;
	
	in.begin(config_in);
	
	// Start I2S out

	auto config_out = out.defaultConfig(TX_MODE);
	config_out.copyFrom(info);
	config_out.i2s_format = I2S_STD_FORMAT;
	config_out.is_master = true;
	config_out.port_no = 1;
	config_out.pin_bck = I2S_BCLK_1;
	config_out.pin_ws = I2S_LRC_1;
	config_out.pin_data = I2S_DOUT_1;
	config_out.pin_mck = I2S_MCLK_1;
	out.begin(config_out);

	effects.begin(config_out);

	mic.begin(effects, converter);

	debugln("Mic Enabled");
	playerControlFlagChanged = LOW;

	displayVolumeRegister();

}	// Close function

/*-----------------------------------------------------------------*/
```

**Comment by pschatzmann** [comment]
I think there is a small bug: StreamCopy mic(out, converter); copies directly to the output w/o effects.
With effects it would need to be StreamCopy mic(effects, converter);

**Reply by cmcooper1971** [reply]
Edited accordingly, thx
